### ISSUE TITLE
Digging doesn't stop swing

### ISSUE THREAD

**Issue by flowl** [PrismarineJS/mineflayer#1631]
Minecraft 1.16.2
Mineflayer 2.41.0

I have a bot digging and after a while the bot teleports to somewhere and uses a chest.
The bug is then, the bot doesn't clear the swing interval:

In lib/plugins/digging.js

    swingInterval = setInterval(() => {
      console.log('XXXX swing')
      bot.swingArm()
    }, 350)

```
Switched bot behavior state to 'mineBlock'.
Digging: With held item
XXXX swing
XXXX  client write
XXXX  finish and clear intervals
XXXX  stopDigging  called
Finished digging: With held item
XXXX  stopDigging  called
XXXX  stopDigging  called
Switched bot behavior state to 'Locate items'.
[FarmBlock] Collect drops
Switched bot behavior state to 'Go to dropped item'.
Switched bot behavior state to 'Find target block'.
XXXX  stopDigging  called
Switched bot behavior state to 'Home: base'. <----- bot chatting /home base
[Chunks] Await Delay <--- the deleport is delayed a few seconds, so sleep
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
[Chunks] Await chunks  
[Debug] Wait for chunks finished <----- teleport finished and chunks finished loading
XXXX  stopDigging  called
Switched bot behavior state to 'Deposit Items'.
Switched bot behavior state to 'Find chest'.
Switched bot behavior state to 'Goto chest'.
XXXX swing
XXXX swing
XXXX swing
...
```

I've put in a lot of `bot.stopDigging()` trying to solve that before looking into mineflayer code.

Although I'm calling `bot.stopDigging()`, it doesn't clear the swinging in the interval function.

```
bot.stopDigging = () => {
    console.log('XXXX  stopDigging  called')
    if (bot.targetDigBlock === null || bot.targetDigBlock === undefined) return
    console.log('XXXX  stopDigging CLEAR')
    bot.removeListener(eventName, onBlockUpdate)
    clearInterval(swingInterval)
    clearTimeout(waitTimeout)
    // ...
```

You can see from the log above,  `stopDigging  CLEAR`  is never logged to the console.
I think the if statement `if (bot.targetDigBlock === null || bot.targetDigBlock === undefined) return` doesn't make sense,
because after digging, the block doesn't exist anymore.

Code for digging:

```
console.log('Digging: ' + label)
// @ts-ignore
this.bot.dig(block).then(() => {
    this.bot.stopDigging()
    console.log('Finished digging: ' + label)
    this.finished = true
}).catch((err) => {
    this.bot.stopDigging()
    console.log('Finished digging with error: ' + label, err)
    delay(500).then(() => {
        this.finished = true
    })
});
```

Notice how `finish and clear intervals` is executed but swing doesn't stop:

```
function finishDigging () {
      console.log('XXXX  finish and clear intervals')
      clearInterval(swingInterval)
      clearTimeout(waitTimeout)
      swingInterval = null
      waitTimeout = null
      if (bot.targetDigBlock !== null && bot.targetDigBlock !== undefined) {
        bot._client.write('block_dig', {
          status: 2, // finish digging
          location: bot.targetDigBlock.position,
          face: 1 // hard coded to always dig from the top
        })
      }
      bot.targetDigBlock = null
      bot.lastDigTime = performance.now()
      bot._updateBlockState(block.position, 0)
    }
```

Here is an example of the log when there is even no swing at all,
which makes me think of a race condition:

```
Switched bot behavior state to 'mineBlock'.
Digging: With held item
XXXX  client write
XXXX  finish and clear intervals
XXXX  stopDigging  called
Finished digging: With held item
XXXX  stopDigging  called
XXXX  stopDigging  called
Switched bot behavior state to 'Locate items'.
Switched bot behavior state to 'Find target block'.
XXXX  stopDigging  called
Switched bot behavior state to 'Home: base'.
[Chunks] Await Delay
[Chunks] Await chunks
[Debug] Wait for chunks finished
XXXX  stopDigging  called
Switched bot behavior state to 'Deposit Items'.
Switched bot behavior state to 'Find chest'.
Switched bot behavior state to 'Goto chest'.
```

**Comment by flowl** [comment]
StackTrace from Possible EventEmitter memory leak:

```
XXXX  client write
XXXX  stopDigging  called
XXXX  stopDigging CLEAR
XXXX  client write
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
XXXX swing
(node:15282) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 diggingAborted listeners added to [EventEmitter]. Use emitter.setMaxListeners() to increase limit
(Use `node --trace-warnings ...` to show where the warning was created)
MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 diggingAborted listeners added to [EventEmitter]. Use emitter.setMaxListeners() to increase limit
    at _addListener (node:events:451:17)
    at EventEmitter.addListener (node:events:467:10)
    at resetPath (/Users/user/Projects/bot/node_modules/mineflayer-pathfinder/index.js:89:11)
    at EventEmitter.<anonymous> (/Users/user/Projects/bot/node_modules/mineflayer-pathfinder/index.js:292:5)
    at EventEmitter.emit (node:events:376:20)
    at addColumn (/Users/user/Projects/bot/node_modules/mineflayer/lib/plugins/blocks.js:111:9)
    at Client.<anonymous> (/Users/user/Projects/bot/node_modules/mineflayer/lib/plugins/blocks.js:327:5)
    at Client.emit (node:events:376:20)
    at FullPacketParser.<anonymous> (/Users/user/Projects/bot/node_modules/minecraft-protocol/src/client.js:91:12)
    at FullPacketParser.emit (node:events:376:20)
(node:15282) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 diggingCompleted listeners added to [EventEmitter]. Use emitter.setMaxListeners() to increase limit
MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 diggingCompleted listeners added to [EventEmitter]. Use emitter.setMaxListeners() to increase limit
    at _addListener (node:events:451:17)
    at EventEmitter.addListener (node:events:467:10)
    at resetPath (/Users/user/Projects/bot/node_modules/mineflayer-pathfinder/index.js:90:11)
    at EventEmitter.<anonymous> (/Users/user/Projects/bot/node_modules/mineflayer-pathfinder/index.js:292:5)
    at EventEmitter.emit (node:events:376:20)
    at addColumn (/Users/user/Projects/bot/node_modules/mineflayer/lib/plugins/blocks.js:111:9)
    at Client.<anonymous> (/Users/user/Projects/bot/node_modules/mineflayer/lib/plugins/blocks.js:327:5)
    at Client.emit (node:events:376:20)
    at FullPacketParser.<anonymous> (/Users/user/Projects/bot/node_modules/minecraft-protocol/src/client.js:91:12)
    at FullPacketParser.emit (node:events:376:20)
```

**Comment by sefirosweb** [comment]
Hi Flow,

I think your issue comes from this.bot.stopDigging()  and teleport instantanly

this.bot.stopDigging() execute a background event for stop, until is not stoppped you must be check is stoped or not, this can take a some physycktick times

you can check if is stopped with 2 events:

```js
// Global variable is setted on moment of start digging
      digging = true
...
      bot.stopDigging()
      bot.on('diggingAborted', detect)
      bot.on('diggingCompleted', detect)

function detect() {
    digging = false
    bot.removeAllListeners('diggingAborted', detect)
    bot.removeAllListeners('diggingCompleted', detect)
  }
```
Caution with calling 2 times to bot.dig(), you MUST be sure the previus dig is stopped (with previus example)
If yo call 2 times on bot.dig() you get a infinite swing

**Comment by flowl** [comment]
Hi @sefirosweb 

thanks for your feedback on the issue.

I get what you mean tho I dont understand how I would even have more than a single `bot.dig()` diggin active:

```
Digging: With held item <---- digging is called once only `bot.dig()`
XXXX swing   <---- bot swing animation
XXXX  client write <---- send stop digging  packet to server
XXXX  finish and clear intervals <----- clearInterval
<----- here it should have a line from     console.log('XXXX  stopDigging CLEAR')  but it doesn't get called as the block is gone?
XXXX  stopDigging  called
Finished digging: With held item <--- digging finished 
XXXX swing <---- the digging stopped but not the swing intervacl
```

The teleporting is delayed with 10 seconds.

My code for digging is based on this source: https://github.com/PrismarineJS/mineflayer-statemachine/blob/master/src/behaviors/behaviorMineBlock.ts
```

**Comment by sefirosweb** [comment]
Hi Flow is a bit hard understand, need more code for help :(

Also I dont understand your stopDigging

```js
this.bot.dig(block).then(() => {
    this.bot.stopDigging() // <-- why you stop? the you use a digblock.then() is already finished
    console.log('Finished digging: ' + label)
    this.finished = true
}).catch((err) => {
    this.bot.stopDigging() // <-- why you stop? the you use a digblock.catch() is already finished but with an error
    console.log('Finished digging with error: ' + label, err)
    delay(500).then(() => {
        this.finished = true
    })
});
```

About https://github.com/PrismarineJS/mineflayer-statemachine/blob/master/src/behaviors/behaviorMineBlock.ts
I think need a bit fixes because when they have error your bot stop all =P

My custom Digging =)
https://github.com/sefirosweb/minecraftLegion/blob/master/BehaviorModules/BehaviorDigBlock.js

**Comment by flowl** [comment]
I made sure to resolve and await everything properly and I even counted how many times dig is called and how many times finished. It's the same number.

I think there is a race condition, if digging is faster than 350ms or something, see the line `            delay(100).then(() => { // The higher this value, the less likely the swing gets stuck!`.


```
private dig(tool, block, label) {
    this.block = block
    return new Promise<void>((resolve, reject) => {
        console.log('[MineBlock] 0000000', this.bot.targetDigBlock)
        this.digStarted++
        this.bot.dig(block).then(() => {
            // this.bot.stopDigging()
            console.log('Finished digging: ' + label)

            // Added delay "graceful stop digging"
            delay(100).then(() => { // The higher this value, the less likely the swing gets stuck!
                console.log('[MineBlock] 1111111', this.bot.targetDigBlock)
                this.digStopped++
                resolve()
            })
        }).catch((err) => {
            console.log('Finished digging with error: ' + label, err)
            delay(1500).then(() => {
                console.log('[MineBlock] 22222222222222222222', this.bot.targetDigBlock)
                this.digStopped++
                reject()
            })
        })
    })
}
```

And the log:
```
[MineBlock] Finished - started digging: 866 stopped digging: 866
Switched bot behavior state to 'Locate items'.
Switched bot behavior state to 'Find target block'.
Switched bot behavior state to 'mineBlock'.
[MineBlock] 0000000 null
Finished digging: With held item: Diamond Shovel
[MineBlock] 1111111 null
[MineBlock] Finished - started digging: 867 stopped digging: 867
Switched bot behavior state to 'Locate items'.
Switched bot behavior state to 'Find target block'.
```

**Comment by flowl** [comment]
**I found the problem!**

My bot does this:


Bot is at 0,64,0
1. Find block to dig -> Pos 0,64,5
2. Go to block (pathfinder.setGoal -> Pos 0,64,5)
3. Dig block at Pos 0,64,5

The race condition is between step 2 and 3.
When the bot is moving to the targeBlock, it sends `dig()` because the block is blocking from reaching the pathfinder goal.
When the bot is close to the targeblock, < 2.7 blocks, the bot is turning over to step 3. With this setup, sometimes the bot is digging twice: one time from pathfinder and one time from my own scripts.

I prevented this in the most simple way by reordering the steps: 1. -> 2. -> again 1.-> 3.

Should digging more than once be prevented generally  in plugins/digging.js per bot instance?
